facebook/contingency-combine.R

```diff
@@ -0,0 +1,182 @@
+  
+  group <- parts[3:length(parts)]
+  # Specify compression format via name, to be parsed by `write_csv` later.
+  partial_name <- paste0(paste0(group, collapse="_"), ".csv.gz")
+
+  return(data.frame(
+    filename=filename,
+    rollup_name=partial_name))
+}
+
+## Helper function to load "seen" file.
+load_seen_file <- function(seen_file) {
+  if (!file.exists(seen_file)) {
+    file.create(seen_file)
+  }
+  
+  seen_files <- readLines(seen_file)
+  return(seen_files)
+}
+
+#' Combine set of input files with existing output file, and save to disk.
+#'
+#' If an input filename has been seen before, the input and output data are
+#' deduplicated to use the newer set of data. Output is saved in gzip-compressed
+#' format.
+#'
+#' @param seen_files Vector of filenames that have been previously loaded into
+#'   an output file.
+#' @param input_dir Directory in which to look for survey CSV files, relative to
+#'   the current working directory.
+#' @param input_files Vector of paths to input files that share a set of
+#'   grouping variables.
+#' @param output_file Path to corresponding output file.
+#' 
+#' @return Character vector of newly-seen filenames.
+combine_and_save_tables <- function(seen_files, input_dir, input_files, output_file) {
+  cols <- cols(
+    .default = col_guess(),
+    survey_geo = col_character(),
+    period_type = col_character(),
+    geo_type = col_character(),
+    aggregation_type = col_character(),
+    country = col_character(),
+    ISO_3 = col_character(),
+    GID_0 = col_character(),
+    region = col_character(),
+    GID_1 = col_character(),
+    state = col_character(),
+    state_fips = col_character(),
+    county = col_character(),
+    county_fips = col_character()
+  )
+  
+  # Get input data.
+  input_df <- map_dfr(
+    file.path(input_dir, input_files),
+    function(f) {
+      read_csv(f, col_types = cols)
+    }
+  )
+  
+  if (file.exists(output_file)) {
+    output_names <- names(read_csv(output_file, n_max = 0L))
+    assert(identical(output_names, names(input_df)),
+           paste0("Column names and/or order differ between new and old input for ", output_file))
+  }
+  
+  # If no input files have been seen before, we can append directly to the
+  # output file without needing to deduplicate. File is created if it doesn't
+  # already exist.
+  any_prev_seen <- any(input_files %in% seen_files)
+  
+  if (!any_prev_seen) {
+    write_csv(input_df, output_file, append=file.exists(output_file))
+  } else {
+    assert(file.exists(output_file),
+           paste0("The output file ", output_file, " does not exist, but non-zero",
+                  " files using the same grouping variables have been seen before."))
+    
+    output_df <- read_csv(output_file, col_types = cols)
+    
+    # Use all columns up to the first "val" column to find unique rows.
+    group_names <- names(output_df)
+    ind_first_val_col <- min(which(startsWith(group_names, "val_")))
```

Comment:
Is it clearly documented in the contingency table generator that the grouping columns must always precede the `val_` columns, and that, say, the standard error can't come before the `val_`, to prevent this from messing up?

Reply:
The `write` function explicitly reorders the columns in a table so that the grouping columns and table metadata (date range, US vs global survey, etc) always come first. I've added more info to the function description to make that clear.

> the standard error can't come before the val_

Good point. I can make the column-finder bit more general so that it only uses up to the first column that's not `val`, `se`, `sample_size`, etc.